# "Human or Robot" 
## Predicting Auction Fraud

On an auction website, human bidders are becoming increasingly frustrated with their inability to win auctions vs. their software-controlled counterparts. As a result, usage from the site's core customer base is plummeting. In order to rebuild customer happiness, the site owners need to eliminate computer generated bidding from their auctions.

The goal of this project is to identify online auction bids that are placed by "robots", thus helping the site owners easily flag these users for removal from their site to prevent unfair auction activity.

<img src="../images/robots-greeting.png" width="500">

## DATA WRANGLING
The process of cleaning and unifying messy and complex datasets for easy access and analysis. This process typically includes manually converting and mapping data from one raw form into another format to allow for more convenient consumption and organization of the data.

**Key Steps:**

- Data Acquisition: Identify and obtain access to the data within sources.

- Joining Data: Combine the edited data for further use and analysis.

- Data Cleansing: Redesign the data into a usable and functional format and correct/remove any bad data.

### Data Description

There are two datasets for this project. One is a bidder dataset that includes a list of bidder information, including their id, payment account, and address. The other is a bid dataset that includes 7.6 million bids on different auctions. The bids in this dataset are all made by mobile devices.

### File descriptions
- train.csv - the training set from the bidder dataset
- test.csv - the test set from the bidder dataset
- bids.csv - the bid dataset

### Data fields

**For the bidder dataset:**

- bidder_id – Unique identifier of a bidder.
- payment_account – Payment account associated with a bidder. These are obfuscated to protect privacy. 
- address – Mailing address of a bidder. These are obfuscated to protect privacy. 
- outcome – Label of a bidder indicating whether or not it is a robot. Value 1.0 indicates a robot, where value 0.0 indicates human. 

The outcome was half hand labeled, half stats-based. There are two types of "bots" with different levels of proof:

1. Bidders who are identified as bots/fraudulent with clear proof. Their accounts were banned by the auction site.

2. Bidder who may have just started their business/clicks or their stats exceed from system wide average. There are no clear proof that they are bots.

**For the bid dataset:**

- bid_id - unique id for this bid
- bidder_id – Unique identifier of a bidder (same as the bidder_id used in train.csv and test.csv)
- auction – Unique identifier of an auction
- merchandise –  The category of the auction site campaign, which means the bidder might come to this site by way of searching for "home goods" but ended up bidding for "sporting goods" - and that leads to this field being "home goods". This categorical field could be a search term, or online advertisement. 
- device – Phone model of a visitor
- time - Time that the bid is made (transformed to protect privacy).
- country - The country that the IP belongs to
- ip – IP address of a bidder (obfuscated to protect privacy).
- url - url where the bidder was referred from (obfuscated to protect privacy).

In [1]:
# import packages
import pandas as pd
import numpy as np

In [2]:
# load datasets
bids_df = pd.read_csv('../large_data/bids.csv')
train = pd.read_csv('../data/train.csv')
test = pd.read_csv('../data/test.csv')

In [3]:
# visualiaze bids_df
bids_df

bid_id                              bidder_id auction  \
0              0  8dac2b259fd1c6d1120e519fb1ac14fbqvax8   ewmzr   
1              1  668d393e858e8126275433046bbd35c6tywop   aeqok   
2              2  aa5f360084278b35d746fa6af3a7a1a5ra3xe   wa00e   
3              3  3939ac3ef7d472a59a9c5f893dd3e39fh9ofi   jefix   
4              4  8393c48eaf4b8fa96886edc7cf27b372dsibi   jefix   
...          ...                                    ...     ...   
7656329  7656329  626159dd6f2228ede002d9f9340f75b7puk8d   3e64w   
7656330  7656330  a318ea333ceee1ba39a494476386136a826dv   xn0y0   
7656331  7656331  f5b2bbad20d1d7ded3ed960393bec0f40u6hn   gja6c   
7656332  7656332  d4bd412590f5106b9d887a43c51b254eldo4f   hmwk8   
7656333  7656333  0ea62aaa9c3ffcc6db584cb69c1f6c4bcripp   c9ox9   

            merchandise    device              time country              ip  \
0               jewelry    phone0  9759243157894736      us   69.166.231.58   
1             furniture    phone1  9759243157894736      in   50.201.125.84   
2            home goods    phone2  9759243157894736      py  112.54.208.157   
3               jewelry    phone4  9759243157894736      in   18.99.175.133   
4               jewelry    phone5  9759243157894736      in    145.138.5.37   
...                 ...       ...               ...     ...             ...   
7656329         jewelry   phone91  9709222052631578      ru  140.204.227.63   
7656330          mobile  phone236  9709222052631578      pl  24.232.159.118   
7656331  sporting goods   phone80  9709222052631578      za   80.237.28.246   
7656332         jewelry  phone349  9709222052631578      my   91.162.27.152   
7656333          mobile   phone82  9709222052631578      jo  160.243.101.60   

                     url  
0        vasstdc27m7nks3  
1        jmqlhflrzwuay9c  
2        vasstdc27m7nks3  
3        vasstdc27m7nks3  
4        vasstdc27m7nks3  
...                  ...  
7656329  cghhmomsaxi6pug  
7656330  wgggpdg2gx5pesn  
7656331  5xgysg14grlersa  
7656332  bhtrek44bzi2wfl  
7656333  vasstdc27m7nks3  

[7656334 rows x 9 columns]

In [4]:
# number of unique bidders
bids_df['bidder_id'].nunique()

6614

In [5]:
# check for missing data
test.isnull().sum().sort_values()

bidder_id          0
payment_account    0
address            0
dtype: int64

In [6]:
# visualize train df
train

bidder_id  \
0     91a3c57b13234af24875c56fb7e2b2f4rb56a   
1     624f258b49e77713fc34034560f93fb3hu3jo   
2     1c5f4fc669099bfbfac515cd26997bd12ruaj   
3     4bee9aba2abda51bf43d639013d6efe12iycd   
4     4ab12bc61c82ddd9c2d65e60555808acqgos1   
...                                     ...   
2008  369515b3af4f8ca582f90271d30b14b6r52aw   
2009  f939c17ffc7c39ac9b35b69e5e75179fv9pe2   
2010  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
2011  0381a69b7a061e9ace2798fd48f1f537mgq57   
2012  84a769adc98498f52debfe57b93a0789556f4   

                            payment_account  \
0     a3d2de7675556553a5f08e4c88d2c228754av   
1     a3d2de7675556553a5f08e4c88d2c228v1sga   
2     a3d2de7675556553a5f08e4c88d2c2280cybl   
3     51d80e233f7b6a7dfdee484a3c120f3b2ita8   
4     a3d2de7675556553a5f08e4c88d2c22857ddh   
...                                     ...   
2008  a1f85275793c4a782f0a668711f41b927ivc9   
2009  a3d2de7675556553a5f08e4c88d2c2286s1m2   
2010  a3d2de7675556553a5f08e4c88d2c22856leq   
2011  fd87037ce0304077079c749f420f0b4c54uo0   
2012  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   

                                    address  outcome  
0     a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
1     ae87054e5a97a8f840a3991d12611fdcrfbq3      0.0  
2     92520288b50f03907041887884ba49c0cl0pd      0.0  
3     4cb9717c8ad7e88a9a284989dd79b98dbevyi      0.0  
4     2a96c3ce94b3be921e0296097b88b56a7x1ji      0.0  
...                                     ...      ...  
2008  e6882cf204a9482edd042b6e31791dfctxzx8      0.0  
2009  b9b03d5a127eb07aeb9163cdcf524e1344ac9      0.0  
2010  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0  
2011  f030a221726fbcdfc4dc7dfd1b381a112hieq      0.0  
2012  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  

[2013 rows x 4 columns]

In [7]:
# number of unique bidders in train data
train['bidder_id'].nunique()

2013

- All variables besides outcome are obfuscated for privacy.
- They are still useful for the model, but it's an issue for interpretability. 

In [8]:
# merge bids_df and train df
auction_df = pd.merge(train, bids_df, on='bidder_id', how='left')
auction_df

bidder_id  \
0        91a3c57b13234af24875c56fb7e2b2f4rb56a   
1        91a3c57b13234af24875c56fb7e2b2f4rb56a   
2        91a3c57b13234af24875c56fb7e2b2f4rb56a   
3        91a3c57b13234af24875c56fb7e2b2f4rb56a   
4        91a3c57b13234af24875c56fb7e2b2f4rb56a   
...                                        ...   
3071248  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
3071249  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
3071250  0381a69b7a061e9ace2798fd48f1f537mgq57   
3071251  84a769adc98498f52debfe57b93a0789556f4   
3071252  84a769adc98498f52debfe57b93a0789556f4   

                               payment_account  \
0        a3d2de7675556553a5f08e4c88d2c228754av   
1        a3d2de7675556553a5f08e4c88d2c228754av   
2        a3d2de7675556553a5f08e4c88d2c228754av   
3        a3d2de7675556553a5f08e4c88d2c228754av   
4        a3d2de7675556553a5f08e4c88d2c228754av   
...                                        ...   
3071248  a3d2de7675556553a5f08e4c88d2c22856leq   
3071249  a3d2de7675556553a5f08e4c88d2c22856leq   
3071250  fd87037ce0304077079c749f420f0b4c54uo0   
3071251  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   
3071252  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   

                                       address  outcome     bid_id auction  \
0        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0    85371.0   yitr4   
1        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   274487.0   btpyy   
2        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   527436.0   kj2ko   
3        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   554135.0   1m5t8   
4        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   607628.0   btpyy   
...                                        ...      ...        ...     ...   
3071248  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0   609792.0   jefix   
3071249  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0  2297899.0   jefix   
3071250  f030a221726fbcdfc4dc7dfd1b381a112hieq      0.0  3400601.0   jqx39   
3071251  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  2007255.0   ntjde   
3071252  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  2007533.0   ntjde   

              merchandise     device          time country               ip  \
0              home goods    phone35  9.759489e+15      ke    82.34.177.248   
1              home goods  phone1004  9.760249e+15      zm   202.130.71.167   
2              home goods     phone4  9.762261e+15      ke     22.54.76.225   
3              home goods     phone4  9.762431e+15      ke    32.13.237.126   
4              home goods     phone4  9.762789e+15      ke  202.236.188.240   
...                   ...        ...           ...     ...              ...   
3071248            mobile   phone316  9.762802e+15      id       28.18.7.49   
3071249            mobile    phone33  9.772601e+15      id  161.237.171.157   
3071250  office equipment    phone11  9.636754e+15      sa     54.212.21.87   
3071251           jewelry    phone59  9.770756e+15      ru       79.72.96.0   
3071252           jewelry    phone59  9.770758e+15      ru       79.72.96.0   

                     url  
0        vasstdc27m7nks3  
1        vasstdc27m7nks3  
2        vasstdc27m7nks3  
3        vasstdc27m7nks3  
4        vasstdc27m7nks3  
...                  ...  
3071248  ar89au37swuzvzz  
3071249  ar89au37swuzvzz  
3071250  cgb17jpn1ixq9hw  
3071251  9ov8hm82p20sxlz  
3071252  9ov8hm82p20sxlz  

[3071253 rows x 12 columns]

In [9]:
# number of unique bidders
auction_df['bidder_id'].nunique()

2013

In [10]:
# number of unique bids
auction_df['bid_id'].nunique()

3071224

In [11]:
# number of unique auctions
auction_df['auction'].nunique()

12740

In [12]:
# check data type
auction_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3071253 entries, 0 to 3071252
Data columns (total 12 columns):
 #   Column           Dtype  
---  ------           -----  
 0   bidder_id        object 
 1   payment_account  object 
 2   address          object 
 3   outcome          float64
 4   bid_id           float64
 5   auction          object 
 6   merchandise      object 
 7   device           object 
 8   time             float64
 9   country          object 
 10  ip               object 
 11  url              object 
dtypes: float64(3), object(9)
memory usage: 304.6+ MB


## DATA CLEANING

In [13]:
# check for missing data
auction_df.isnull().sum().sort_values()

bidder_id             0
payment_account       0
address               0
outcome               0
bid_id               29
auction              29
merchandise          29
device               29
time                 29
ip                   29
url                  29
country            2730
dtype: int64

- There are 29 missing data points for several features coming from the train dataset. Let's take a look at it!


In [14]:
# create an array with bidder_id based on missing bid_id
no_bids = auction_df[pd.isnull(auction_df['bid_id'])]['bidder_id'].unique()
len(no_bids)
#no_bids

29

In [15]:
# check for outcome of each missing bid_id
auction_df[auction_df['bidder_id'].isin(no_bids)]['outcome']

17583      0.0
42424      0.0
129439     0.0
214915     0.0
221776     0.0
226432     0.0
255107     0.0
305088     0.0
333331     0.0
409451     0.0
654549     0.0
657208     0.0
788183     0.0
790914     0.0
794927     0.0
810369     0.0
883351     0.0
1018748    0.0
1019599    0.0
1021349    0.0
1123023    0.0
1139921    0.0
1217466    0.0
1465813    0.0
1482542    0.0
1648201    0.0
1974826    0.0
2139077    0.0
2999380    0.0
Name: outcome, dtype: float64

- All the 29 bidders are labeled as humans, since we are interested in detecting robot, we can drop these.

In [16]:
# drop 29 missing values
auction_df = auction_df[~auction_df['bidder_id'].isin(no_bids)]
auction_df

bidder_id  \
0        91a3c57b13234af24875c56fb7e2b2f4rb56a   
1        91a3c57b13234af24875c56fb7e2b2f4rb56a   
2        91a3c57b13234af24875c56fb7e2b2f4rb56a   
3        91a3c57b13234af24875c56fb7e2b2f4rb56a   
4        91a3c57b13234af24875c56fb7e2b2f4rb56a   
...                                        ...   
3071248  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
3071249  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
3071250  0381a69b7a061e9ace2798fd48f1f537mgq57   
3071251  84a769adc98498f52debfe57b93a0789556f4   
3071252  84a769adc98498f52debfe57b93a0789556f4   

                               payment_account  \
0        a3d2de7675556553a5f08e4c88d2c228754av   
1        a3d2de7675556553a5f08e4c88d2c228754av   
2        a3d2de7675556553a5f08e4c88d2c228754av   
3        a3d2de7675556553a5f08e4c88d2c228754av   
4        a3d2de7675556553a5f08e4c88d2c228754av   
...                                        ...   
3071248  a3d2de7675556553a5f08e4c88d2c22856leq   
3071249  a3d2de7675556553a5f08e4c88d2c22856leq   
3071250  fd87037ce0304077079c749f420f0b4c54uo0   
3071251  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   
3071252  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   

                                       address  outcome     bid_id auction  \
0        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0    85371.0   yitr4   
1        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   274487.0   btpyy   
2        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   527436.0   kj2ko   
3        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   554135.0   1m5t8   
4        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0   607628.0   btpyy   
...                                        ...      ...        ...     ...   
3071248  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0   609792.0   jefix   
3071249  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0  2297899.0   jefix   
3071250  f030a221726fbcdfc4dc7dfd1b381a112hieq      0.0  3400601.0   jqx39   
3071251  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  2007255.0   ntjde   
3071252  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  2007533.0   ntjde   

              merchandise     device          time country               ip  \
0              home goods    phone35  9.759489e+15      ke    82.34.177.248   
1              home goods  phone1004  9.760249e+15      zm   202.130.71.167   
2              home goods     phone4  9.762261e+15      ke     22.54.76.225   
3              home goods     phone4  9.762431e+15      ke    32.13.237.126   
4              home goods     phone4  9.762789e+15      ke  202.236.188.240   
...                   ...        ...           ...     ...              ...   
3071248            mobile   phone316  9.762802e+15      id       28.18.7.49   
3071249            mobile    phone33  9.772601e+15      id  161.237.171.157   
3071250  office equipment    phone11  9.636754e+15      sa     54.212.21.87   
3071251           jewelry    phone59  9.770756e+15      ru       79.72.96.0   
3071252           jewelry    phone59  9.770758e+15      ru       79.72.96.0   

                     url  
0        vasstdc27m7nks3  
1        vasstdc27m7nks3  
2        vasstdc27m7nks3  
3        vasstdc27m7nks3  
4        vasstdc27m7nks3  
...                  ...  
3071248  ar89au37swuzvzz  
3071249  ar89au37swuzvzz  
3071250  cgb17jpn1ixq9hw  
3071251  9ov8hm82p20sxlz  
3071252  9ov8hm82p20sxlz  

[3071224 rows x 12 columns]

In [17]:
# missing values for country
missing = pd.isnull(auction_df['country']).sum()/len(auction_df)*100
print(f" %0.2f %% observations missing country entry." %
      float(missing))

 0.09 % observations missing country entry.


In [18]:
# create an array with bidder_id based on missing country
no_bids = auction_df[pd.isnull(auction_df['country'])]['bidder_id'].unique()
len(no_bids)
#no_bids

184

In [19]:
# check for number of unique bidder_id 
auction_df['bidder_id'].nunique()

1984

- There are only 0.09% of values missing from the entire dataset. However, it represents 184 out of 1984 total bidders (29 were already dropped). So, we'll impute missing values for country with 'other', in order to preserve information regarding unique bidder_ids. And, it differentiates original from imputed data.
- We could also use KNN to impute country values based on clusters.

In [20]:
# just checking most frequent country
auction_df['country'].mode()

0    in
dtype: object

In [21]:
# impute missing values for country column with 'other'
auction_df = auction_df.replace(np.nan, 'other')

In [22]:
# check for missing data
auction_df.isnull().sum().sort_values()

bidder_id          0
payment_account    0
address            0
outcome            0
bid_id             0
auction            0
merchandise        0
device             0
time               0
country            0
ip                 0
url                0
dtype: int64

In [23]:
# check for duplicates
auction_df.duplicated().sum()

0

In [24]:
# reorganize columns order for readibility
auction_df = auction_df[['bid_id', 'auction', 
                         'merchandise', 'device', 'time', 
                         'country', 'ip', 'url', 
                         'bidder_id', 'payment_account', 
                         'address', 'outcome']]

auction_df

bid_id auction       merchandise     device          time country  \
0          85371.0   yitr4        home goods    phone35  9.759489e+15      ke   
1         274487.0   btpyy        home goods  phone1004  9.760249e+15      zm   
2         527436.0   kj2ko        home goods     phone4  9.762261e+15      ke   
3         554135.0   1m5t8        home goods     phone4  9.762431e+15      ke   
4         607628.0   btpyy        home goods     phone4  9.762789e+15      ke   
...            ...     ...               ...        ...           ...     ...   
3071248   609792.0   jefix            mobile   phone316  9.762802e+15      id   
3071249  2297899.0   jefix            mobile    phone33  9.772601e+15      id   
3071250  3400601.0   jqx39  office equipment    phone11  9.636754e+15      sa   
3071251  2007255.0   ntjde           jewelry    phone59  9.770756e+15      ru   
3071252  2007533.0   ntjde           jewelry    phone59  9.770758e+15      ru   

                      ip              url  \
0          82.34.177.248  vasstdc27m7nks3   
1         202.130.71.167  vasstdc27m7nks3   
2           22.54.76.225  vasstdc27m7nks3   
3          32.13.237.126  vasstdc27m7nks3   
4        202.236.188.240  vasstdc27m7nks3   
...                  ...              ...   
3071248       28.18.7.49  ar89au37swuzvzz   
3071249  161.237.171.157  ar89au37swuzvzz   
3071250     54.212.21.87  cgb17jpn1ixq9hw   
3071251       79.72.96.0  9ov8hm82p20sxlz   
3071252       79.72.96.0  9ov8hm82p20sxlz   

                                     bidder_id  \
0        91a3c57b13234af24875c56fb7e2b2f4rb56a   
1        91a3c57b13234af24875c56fb7e2b2f4rb56a   
2        91a3c57b13234af24875c56fb7e2b2f4rb56a   
3        91a3c57b13234af24875c56fb7e2b2f4rb56a   
4        91a3c57b13234af24875c56fb7e2b2f4rb56a   
...                                        ...   
3071248  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
3071249  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
3071250  0381a69b7a061e9ace2798fd48f1f537mgq57   
3071251  84a769adc98498f52debfe57b93a0789556f4   
3071252  84a769adc98498f52debfe57b93a0789556f4   

                               payment_account  \
0        a3d2de7675556553a5f08e4c88d2c228754av   
1        a3d2de7675556553a5f08e4c88d2c228754av   
2        a3d2de7675556553a5f08e4c88d2c228754av   
3        a3d2de7675556553a5f08e4c88d2c228754av   
4        a3d2de7675556553a5f08e4c88d2c228754av   
...                                        ...   
3071248  a3d2de7675556553a5f08e4c88d2c22856leq   
3071249  a3d2de7675556553a5f08e4c88d2c22856leq   
3071250  fd87037ce0304077079c749f420f0b4c54uo0   
3071251  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   
3071252  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   

                                       address  outcome  
0        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
1        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
2        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
3        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
4        a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
...                                        ...      ...  
3071248  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0  
3071249  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0  
3071250  f030a221726fbcdfc4dc7dfd1b381a112hieq      0.0  
3071251  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  
3071252  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  

[3071224 rows x 12 columns]

In [25]:
# save dataset to csv
#auction_df.to_csv("auction_df", index=False)

--------